src/Microsoft.Health.Fhir.CosmosDb/Features/Health/CosmosHealthCheck.cs

```diff
@@ -80,36 +80,58 @@ public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context
                     {
-                        _logger.LogWarning(coce, "Failed to connect to the data store. External cancellation requested.");
-                        return HealthCheckResult.Unhealthy("Failed to connect to the data store. External cancellation requested.");
+                        const string message = "Failed to connect to the data store. External cancellation requested.";
+
+                        // Handling an extenal cancellation.
+                        // No reasons to retry as the cancellation was external to the health check.
+
+                        _logger.LogWarning(coce, message);
+                        return HealthCheckResult.Unhealthy(message);
                     }
                     else if (attempt >= maxNumberAttempts)
                     {
-                        _logger.LogWarning(coce, "Failed to connect to the data store. There were {NumberOfAttempts} attempts to connect to the data store, but they suffered a '{ExceptionType}'.", attempt, nameof(CosmosOperationCanceledException));
+                        // This is a very rare situation. This condition indicates that multiple attempts to connect to the data store happened, but they were not successful.
+
+                        _logger.LogWarning(
+                            coce,
+                            "Failed to connect to the data store. There were {NumberOfAttempts} attempts to connect to the data store, but they suffered a '{ExceptionType}'.",
+                            attempt,
+                            nameof(CosmosOperationCanceledException));
                         return HealthCheckResult.Unhealthy("Failed to connect to the data store. Operation canceled.");
                     }
                     else
                     {
                         // Number of attempts not reached. Allow retry.
+
                         _logger.LogWarning(coce, "Failed to connect to the data store. Attempt {NumberOfAttempts}. '{ExceptionType}'.", attempt, nameof(CosmosOperationCanceledException));
                     }
                 }
                 catch (CosmosException ex) when (ex.IsCmkClientError())
                 {
-                    return HealthCheckResult.Unhealthy(
-                        "Connection to the data store was unsuccesful because the client's customer-managed key is not available.",
+                    // Handling CMK errors.
+
+                    _logger.LogWarning(
+                        ex,
+                        "Connection to the data store was unsuccesful because the client's customer-managed key is not available.");
+
+                    return HealthCheckResult.Degraded(
+                        "Failed to connect to the data store. Customer-managed key is not available.",
                         exception: ex,
                         new Dictionary<string, object>() { { "IsCustomerManagedKeyError", true } });
```

Comment:
Have you considered using this enum? https://github.com/microsoft/healthcare-shared-components/blob/main/src/Microsoft.Health.Core/Features/Health/HealthStatusReason.cs

Dicom uses this in our monitor to figure out what status to report in the health state. DICOM and FHIR don't have to use the same ones, but if you want to share common things like the metric publisher, it will make it easier to 
https://github.com/microsoft/healthcare-shared-components/blob/main/src/Microsoft.Health.Core/Features/HealthPublisher/HealthCheckMetricPublisher.cs#L51

Reply:
Plus, thats how I have it in SQL health check (which I'm assuming FHIR uses too?) so it makes sense to stay consistent https://github.com/microsoft/healthcare-shared-components/blob/main/src/Microsoft.Health.SqlServer/Features/Health/SqlServerHealthCheck.cs#L51

Reply:
Thanks for the suggestion! And I like it. 

Besides the facts that DICOM is DICOM, and FHIR is FHIR, I like the idea of standardizing external behaviors. Thanks!

Reply:
PR updated. Let me know if it makes sense.